### ISSUE TITLE
--coverage seems to cause misses when building in multiple directories while using CCACHE_NOHASHDIR

### ISSUE THREAD

**Issue by yaahallo** [ccache/ccache#268]
Hi I'm trying to setup ccache to work with our build automation where each compilation is given a new build id and is built in a directory called /build/automated/$buildid/src

if building without debug information it seems to work as is, if building with debug information, specifically `-ggdb3` it requires either CACHE_NOHASHDIR=1 or -fdebug-prefix-map=$(pwd)=.

But our primary use case that I'm trying to speed up also uses --coverage. When rebuilding in the same directory the --coverage flag it works fine, but as soon as I introduced two directories and --coverage it gets cache misses.

below are some logs showing how to reproduce the issue

the first file is a single run without the coverage flag; it clears the cache before running, showing a miss in the first dir then a hit in the second. 

The second file is a single run which clears the cache and builds in each dir with -ggdb3 and --coverage, and it misses in both dirs,

the 3rd file is a single run after the previous one, without clearing the cache, and it shows both directories getting a cache hit

[singlerun-with-ggdb3.log](https://github.com/ccache/ccache/files/2086465/singlerun-with-ggdb3.log)
[first-run-with-coverage.log](https://github.com/ccache/ccache/files/2086466/first-run-with-coverage.log)
[second-run-with-coverage.log](https://github.com/ccache/ccache/files/2086467/second-run-with-coverage.log)

I was able to reproduce the same issue using debug-prefix-map, here's the alternate version

these files are equivalent to the first and second files above.

[prefixmap-withoutcov.log](https://github.com/ccache/ccache/files/2086476/prefixmap-withoutcov.log)
[prefixmap-withcov.log](https://github.com/ccache/ccache/files/2086477/prefixmap-withcov.log)

In each file, I show the log output of my test-ccache.sh script, followed by (in some files) the content of hello.cpp, and after that (in all files) the content of test-ccache.sh.

please let me know if you need any additional information beyond what is in the log files.

**Comment by jrlusby** [comment]
looks like it might actually have been this issue. https://github.com/ccache/ccache/issues/219, setting -fprofile-dir lets me get hits, I'm now checking how that affects coverage generation and maybe debugging all of that. Its cool if this is a no fix but I think maybe its worth adding to the compiling in different directories

**Comment by afbjorklund** [comment]
We need some better documentation, on how debuginfo and coverage affects the hashing. It could also need better debugging support, as with issue #226

**Comment by afbjorklund** [comment]
These are the relevant entries from your logs:
`[2018-06-08T16:51:52.436823 31589] Hashing coverage path /home/jlusby/git/dir1/hello.gcda`
`[2018-06-08T16:52:51.526162 31778] Hashing coverage path /home/jlusby/git/dir2/hello.gcda`

Since those paths are in the generated .o files.

**Comment by afbjorklund** [comment]
I think the main confusion comes from the fact that `hash_dir` *only* affects the debuginfo and the _source_ directory, so that is looking for "generating_debuginfo" and `-fdebug-prefix-map`
But it currently does *not* affect the coverage and the _output_ directory, which is looking for "generating_coverage" and `-fprofile-dir`. That might not be obvious from the current manual...

The main hack here is the legacy usage of hash_dir=false, because that is trying to re-use cache objects which are pointing to stale debuginfo (source path) and stale gcovdata (output path) which is "dangerous".
It is much better to encourage people to generate reusable cache objects, by changing the absolute paths that are (unfortunately) encoded into those object files. Even better would be if they were relative...

**Comment by jrosdahl** [comment]
@yaahallo: Are you OK with @afbjorklund's answers or do you have any outstanding questions?

**Comment by jrosdahl** [comment]
> That might not be obvious from the current manual...

@afbjorklund: Would you like to work on this or at least open an issue about it?

**Comment by afbjorklund** [comment]
Opened #294

**Comment by yaahallo** [comment]
@jrosdahl I think so, its been a while since I looked at this. I do know that I got it working though we decided to just disable ccache for automated builds that include coverage because we'd have to reexport information to convert the relative paths in the cached files into absolute paths and a bunch of other shit that ended up causing too much change for it to be desired. So yea no outstanding questions.
